### PR TITLE
Problem: Stale worker documents present in the db

### PR DESCRIPTION
**Pull request by dkliban**
Solution: A custom queryset for the Worker model allows filtering out
worker records which have not been updated in more than 25 seconds.
This queryset is used in two places:

- Status API

- Resource manager code that looks for available workers

closes #2496
https://pulp.plan.io/issues/2496

### CODE REVIEW

File: server/pulp/server/db/querysets.py [hunk 1]
```diff
@@ -116,6 +119,27 @@ def get_or_404(self, **kwargs):
             raise pulp_exceptions.MissingResource(**kwargs)
 
 
+class WorkerQuerySet(CriteriaQuerySet):
+    """
+    Custom queryset for workers
+    """
+
+    def get_workers_online(self):
```
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: What about naming this `get_online()` or just `online()`?

If changed it would read as either `Worker.objects.get_online()` or `Worker.objects.online()`

Comment (dkliban) [inline server/pulp/server/db/querysets.py]: I prefer get_online(). I'll update the PR with that.

File: server/pulp/server/db/querysets.py [hunk 2]
```diff
@@ -116,6 +119,27 @@ def get_or_404(self, **kwargs):
...
+        """
+        Returns a queryset that contains only worker documents that have been updated in the last
```
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: The first line of a multiline docstring should be a single line. Please shorten this to 1 line, or add a shorter 1 line above with a whitespace between. See the PEP here:  https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: s/worker/Worker/

File: server/pulp/server/db/querysets.py (hunk 1 above)
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: thx

File: server/pulp/server/db/querysets.py [hunk 3]
```diff
@@ -116,6 +119,27 @@ def get_or_404(self, **kwargs):
...
+        25 seconds.
+
+        :return: mongoengine queryset object
+        :rtype:  mongoengine.queryset.QuerySet
+        """
+        query_set = self
```
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: This read strangely to me. I think having L#138 be `self.filter....` would be fine. This is not a required change.

Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: Actually, maybe this is better and should be kept. It is explicit.

**Comment by bmbouter** [comment]
The commit message title and body are too long for the guidelines.

http://docs.pulpproject.org/en/2.12/nightly/dev-guide/contributing/branching.html#commit-messages

**Comment by bmbouter** [comment]
This has test failure which need to be fixed.

**Comment by dkliban** [comment]
ok test

**Comment by dkliban** [comment]
ok test

File: server/pulp/server/db/querysets.py [hunk 4]
```diff
@@ -116,6 +119,28 @@ def get_or_404(self, **kwargs):
...
+class WorkerQuerySet(CriteriaQuerySet):
+    """
+    Custom queryset for workers
+    """
+
+    def get_online(self):
+        """
+        Returns a queryset with a subset of Worker documents.
+
+        The queryset is filtered to remove any Worker document that have not been updated in the
```
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: s/have/has/

Comment (dkliban) [inline server/pulp/server/db/querysets.py]: any [one]  ... i'll push the change shortly.

File: server/pulp/server/db/querysets.py [hunk 5]
```diff
@@ -116,6 +119,28 @@ def get_or_404(self, **kwargs):
...
+        The queryset is filtered to remove any Worker document that has not been updated in the
+        last 25 seconds.
+
+        :return: mongoengine queryset object
+        :rtype:  mongoengine.queryset.QuerySet
+        """
+        query_set = self
+        now = ensure_tz(datetime.utcnow())
+        oldest_heartbeat_time = now - timedelta(seconds=constants.PULP_PROCESS_TIMEOUT_INTERVAL)
+        query_set.filter(last_heartbeat__gte=oldest_heartbeat_time)
```
Comment (bmbouter) [inline server/pulp/server/db/querysets.py]: The return value of this statement needs to be returned. As it's written now the original queryset is returned. Remove L#141 and make this `return query_set.filter(last_heartbeat__gte=oldest_heartbeat_time)`
